fdbserver/DataDistribution.actor.cpp

```diff
@@ -1360,6 +1391,418 @@ ACTOR Future<Void> bulkLoadEngineCore(Reference<DataDistributor> self, Future<Vo
+	state Database cx = self->txnProcessor->context();
+	state Transaction tr(cx);
+	state BulkLoadTaskState bulkLoadTask(bulkLoadJob.getJobId(), bulkLoadJob.getManifest());
+	loop {
+		try {
+			// At any time, there must be at most one bulkload job
+			wait(checkMoveKeysLock(&tr, self->context->lock, self->context->ddEnabledState.get()));
+			bool doUpdate = wait(updateBulkLoadJobMetadata(&tr, bulkLoadJob));
+			if (!doUpdate) {
+				TraceEvent(SevVerbose, "BulkLoadJobTriggerTaskNotUpdate", self->ddId)
+				    .detail("BulkLoadJob", bulkLoadJob.toString());
+			}
+			wait(setBulkLoadSubmissionTransaction(&tr, bulkLoadTask));
+			// setBulkLoadSubmissionTransaction shuts down traffic to the range
+			wait(tr.commit());
+			break;
+		} catch (Error& e) {
+			wait(tr.onError(e));
+		}
+	}
+	return bulkLoadTask;
+}
+
+ACTOR Future<BulkLoadTaskState> bulkLoadJobGetOngoingBulkLoadTask(Reference<DataDistributor> self, KeyRange range) {
+	state Database cx = self->txnProcessor->context();
+	state std::vector<BulkLoadTaskState> existTasks = wait(getValidBulkLoadTasksWithinRange(cx, range));
+	ASSERT(existTasks.empty() || existTasks.size() == 1);
+	if (existTasks.empty()) {
+		TraceEvent(SevWarn, "DDBulkLoadJobRunOnTaskOutdated", self->ddId)
+		    .setMaxEventLength(-1)
+		    .setMaxFieldLength(-1)
+		    .detail("Location", "Read bulkload");
+		throw bulkload_task_outdated();
+	}
+	return existTasks[0];
+}
+
+ACTOR Future<Void> bulkLoadJobWaitUntilBulkLoadTaskComplete(Reference<DataDistributor> self,
+                                                            BulkLoadTaskState bulkLoadTask) {
+	ASSERT(bulkLoadTask.isValid());
+	state Database cx = self->txnProcessor->context();
+	state Transaction tr(cx);
+	loop {
+		try {
+			wait(store(bulkLoadTask,
+			           getBulkLoadTask(&tr,
+			                           bulkLoadTask.getRange(),
+			                           bulkLoadTask.getTaskId(),
+			                           { BulkLoadPhase::Submitted,
+			                             BulkLoadPhase::Triggered,
+			                             BulkLoadPhase::Running,
+			                             BulkLoadPhase::Complete,
+			                             BulkLoadPhase::Acknowledged })));
+			if (bulkLoadTask.phase == BulkLoadPhase::Complete || bulkLoadTask.phase == BulkLoadPhase::Acknowledged) {
+				break;
+			}
+		} catch (Error& e) {
+			wait(tr.onError(e));
+		}
+		wait(delay(10.0));
+	}
+	return Void();
+}
+
+ACTOR Future<Void> bulkLoadJobFinalizeBulkLoadTask(Reference<DataDistributor> self,
+                                                   BulkLoadTaskState bulkLoadTask,
+                                                   BulkLoadJobState bulkLoadJob) {
+	state Database cx = self->txnProcessor->context();
+	state Transaction tr(cx);
+	loop {
+		try {
+			wait(checkMoveKeysLock(&tr, self->context->lock, self->context->ddEnabledState.get()));
+			wait(setBulkLoadAcknowledgeTransaction(&tr, bulkLoadTask.getRange(), bulkLoadTask.getTaskId()));
```

Comment:
I see how setBulkLoadAckTrans is used now. Can the method not be setBulkLoadFinalizeTransaction?

Reply:
Here we use the "Ack" in the method name because we want to get the name aligned to the bulkload task phase set by this method. This method is not used by users to we may not want to hide this detail in the name.